sql/lantern.sql

```diff
@@ -193,3 +195,223 @@ BEGIN
     END LOOP;
 END $$ LANGUAGE plpgsql VOLATILE;
+
+-------------------------------------
+-------- Product Quantization -------
+-------------------------------------
+
+CREATE FUNCTION ldb_pqvec_in(cstring, oid, integer) RETURNS pqvec AS 'MODULE_PATHNAME' LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+CREATE FUNCTION ldb_pqvec_out(pqvec) RETURNS cstring AS 'MODULE_PATHNAME' LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+CREATE FUNCTION ldb_pqvec_recv(internal, oid, integer) RETURNS pqvec AS 'MODULE_PATHNAME' LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+CREATE FUNCTION ldb_pqvec_send(pqvec) RETURNS bytea AS 'MODULE_PATHNAME' LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+CREATE FUNCTION ldb_cast_array_pqvec(int[], integer, boolean) RETURNS pqvec	AS 'MODULE_PATHNAME' LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+CREATE FUNCTION ldb_cast_pqvec_array(pqvec, integer, boolean) RETURNS int[]	AS 'MODULE_PATHNAME' LANGUAGE C IMMUTABLE STRICT PARALLEL SAFE;
+
+CREATE TYPE pqvec (
+	INPUT     = ldb_pqvec_in,
+	OUTPUT    = ldb_pqvec_out,
+	RECEIVE   = ldb_pqvec_recv,
+	SEND      = ldb_pqvec_send,
+	STORAGE   = extended
+);
+
+CREATE CAST (integer[] AS pqvec)
+	WITH FUNCTION ldb_cast_array_pqvec(integer[], integer, boolean) AS ASSIGNMENT;
+	
+CREATE CAST (pqvec AS integer[])
+	WITH FUNCTION ldb_cast_pqvec_array(pqvec, integer, boolean) AS ASSIGNMENT;
+	
+CREATE OR REPLACE FUNCTION create_codebook(tbl REGCLASS, col NAME, cluster_cnt INT, subset_count INT, distance_metric TEXT)
+RETURNS NAME AS $$
+DECLARE
+  stmt TEXT;
+  res REAL[];
+  codebooks REAL[][][];
+  i INT;
+  subset_len INT;
+  end_idx INT;
+  codebook_table NAME;
+  dim INT;
+BEGIN
+  
+  stmt := format('SELECT array_length(%I, 1) FROM %I WHERE %1$I IS NOT NULL LIMIT 1', col, tbl);
+  EXECUTE stmt INTO dim;
+
+	-- Get codebooks
+	codebooks := _lantern_internal.create_pq_codebook(tbl, col, cluster_cnt, subset_count, distance_metric);
+
+	-- Create codebook table
+  codebook_table := format('_lantern_codebook_%s', tbl);
+  stmt := format('DROP TABLE IF EXISTS %s CASCADE', codebook_table);
+  EXECUTE stmt;
+  
+  stmt:= format('CREATE TABLE %s(subvector_id INT, centroid_id INT, c REAL[]);', codebook_table);
+  EXECUTE stmt;
+  
+  stmt:= format('CREATE INDEX ON %s USING BTREE(subvector_id, centroid_id);', codebook_table);
+  EXECUTE stmt;
+  
+  -- Iterate over codebooks and insert into table
+  FOR i IN 1..subset_count loop
+  	FOR k IN 1..cluster_cnt loop
+  	  -- centroid_id is k-1 because k is in range[0,255] but postgres arrays start from index 1
+      stmt := format('INSERT INTO %I(subvector_id, centroid_id, c) VALUES (%s, %s, ARRAY(SELECT * FROM unnest(''%s''::REAL[])))', codebook_table, i - 1, k - 1, codebooks[i:i][k:k]);
+      EXECUTE stmt;
+  	END LOOP;
+  END LOOP;
+
+  return codebook_table;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Compress vector using codebook
+CREATE OR REPLACE FUNCTION _lantern_internal.compress_vector(v REAL[], subset_count INTEGER, codebook regclass, distance_metric TEXT)
```

Comment:
does this need to take `subset_count`?
can't it infer the value cheaply from the codebook?

Reply:
This one is internal function, theres external wrapper which does infer the value from codebook